LFS-533: Allow upload of vocabulary files

IntegralProgrammer opened the pull request:
Allows for vocabularies to be uploaded in RDF/XML format, thus allowing for the installation of vocabularies that are not available through BioPortal.


To test:

1. Build the `LFS-533_test` branch
2. Download the **Human Phenotype Ontology** in **RDF/XML** format from https://bioportal.bioontology.org/ontologies/HP
3. Ensure that the `BIOPORTAL_APIKEY` environment variable is **not** set
4. Start LFS
5. Go to *Administration* ... *Vocabularies* and install the downloaded ontology file with the following parameters; `Identifier` = `HP`, `Name` = `Human Phenotype Ontology`, `Version` = `1`
6. Once installation is finished, create a new *Demographics* form.
7. Type something like `myopia` in the co-morbidities field. Auto-complete should happen and answers should allow for selection.

IntegralProgrammer committed:
LFS-533: Allow upload of vocabulary files

sdumitriu commented on modules/vocabularies/src/main/java/ca/sickkids/ccm/lfs/vocabularies/internal/FileUploadIndexer.java:
```diff
@@ -0,0 +1,275 @@
+     * @param term the term data
+     * @throws VocabularyIndexException when a node cannot be created
+     */
+    private void createVocabularyTermNode(VocabularyTermSource term)
```
This looks like a method that should be shared between the different indexers. Can you refactor them toward that goal?

IntegralProgrammer replied:
Done in commit `4d93f38c` :heavy_check_mark:

sashaandjic commented:
tested by following the steps described above
![image](https://user-images.githubusercontent.com/23243244/93424162-737afc00-f885-11ea-821d-c7341801d256.png)
![image](https://user-images.githubusercontent.com/23243244/93424172-7a097380-f885-11ea-9254-f1c98a6bb5cf.png)

sashaandjic commented:
I was able to install GO and HANCESTRO with the method described above. For some reason I could not get HANCESTRO work in the **Race/Ethnicity reported** field. It worked well in **Co-morbidities** field as you can see in the screenshots below. 
![image](https://user-images.githubusercontent.com/23243244/93822417-353b5f00-fc2e-11ea-92fb-c1da81595c87.png)
![image](https://user-images.githubusercontent.com/23243244/93822570-6caa0b80-fc2e-11ea-9f13-f20f121dbac2.png)
![image](https://user-images.githubusercontent.com/23243244/93822681-93684200-fc2e-11ea-89e7-a2d73438f245.png)

I was not able to have more than one vocabulary linked to one field.

marta- commented:
> I was not able to have more than one vocabulary linked to one field.

That's expected, LFS-198 isn't merged yet.

marta- commented:
> For some reason I could not get HANCESTRO work in the **Race/Ethnicity reported** field.

This is out of scope, but requires further investigation. The only notable difference I see at first glance is that Race/Ethnicity has display mode list+input and has predefined options, while Co-morbidities  has display mode input.

@sashaandjic can you do a quick test and change the display mode for Race/ethnicity to see if the suggestions work then?

sashaandjic commented:
Changing display mode to input didn't work.
I the meantime I installed ORDO and it worked well in co-morbidities.
![image](https://user-images.githubusercontent.com/23243244/93824312-3cb03780-fc31-11ea-979f-3938206c0ec6.png)

sashaandjic commented:
I cannot believe I missed this. Once I removed vocabularyFilter from Race/Ethnicity field it WORKED. Duh!

sashaandjic reviewed (approved):
Tested.
Succesfully uploaded HP, GO, HANCESTRO, ORDO and CHEBI voabularies.

sdumitriu reviewed (changes requested):
Problems:

- The INSTALL button is always clickable, and clicking it will actually send a request, and a new row will be added to the INSTALLED table, even though nothing is installed
- After installing HP, it will show up in the INSTALLED table, but with an "Install" button instead of the "Uninstall" one
- Why are the inputs doubled, once as a hidden input (with a proper `name` attribute) and once as a unnamed MUI input? This actually causes a JS console warning about inputs with a value but no onchange prop
- The indexing servlet returns a 200 status even if there's an error, but with an error payload. The response should be a 500, and the frontend should handle the error payload.
- Why only `OWL` is supported? The same code can handle all other formats, the only thing that is preventing it now is line 130 where we hadcode the format. Is there a way to detect the format from the input filename?
- Also see my previous comment about possible code duplication in `createVocabularyTermNode`

sdumitriu commented:
And out of scope for this PR, but the INSTALLED table should not depend so much on the bioportal service, it should be able to display everything it does so far, except the Upgrade button.

IntegralProgrammer committed:
Return HTTP 500 (not HTTP 200) when vocabulary parsing fails

IntegralProgrammer committed:
Display a 'Failed' button if server fails to load vocabulary

Properly disable buttons when vocabulary is installing

Change button from 'Select OWL' to 'OWL Selected' to notify user that an OWL file has been selected

Change color scheme for sucessfully installed vocabulary button

IntegralProgrammer committed:
When vocabularies are installed from a local file source, they are automatically set into the Latest phase so that an Uninstall button is always present

Vocabularies can now be installed from local files even if a BioPortal API key is not present

IntegralProgrammer committed:
Fix JS console warning about inputs with a value but no onchange prop

IntegralProgrammer committed:
Refactor: Move getSourceFormat() from EbiRepositoryHandler to the utility class OntologyFormatDetection

IntegralProgrammer committed:
Use OntologyFormatDetection.getSourceFormat() to determine the ontology format from the file extension of the uploaded ontology file

IntegralProgrammer committed:
Refactor: Move createVocabularyTermNode() to the OntologyIndexerUtils utility class

IntegralProgrammer committed:
Refactor: Move createVocabularyTermNode() to the OntologyIndexerUtils utility class

IntegralProgrammer committed:
More refactoring: Move createVocabularyNode() and saveSession() into OntologyIndexerUtils

IntegralProgrammer committed:
Fix bug where Uninstall/About buttons disappeared upon page refresh

IntegralProgrammer committed:
Change 'Install from OWL file' message to 'Install from local file'

IntegralProgrammer committed:
Change enctype to encType to fix browser console warning

IntegralProgrammer commented:
>     * The INSTALL button is always clickable, and clicking it will actually send a request, and a new row will be added to the INSTALLED table, even though nothing is installed
> 

Fixed by commit `b1b927772` :heavy_check_mark:

IntegralProgrammer commented:
> 
>     * After installing HP, it will show up in the INSTALLED table, but with an "Install" button instead of the "Uninstall" one
> 
Fixed by commit `6a43fde3ef` :heavy_check_mark:

IntegralProgrammer commented:
> 
>     * Why are the inputs doubled, once as a hidden input (with a proper `name` attribute) and once as a unnamed MUI input? This actually causes a JS console warning about inputs with a value but no onchange prop
> 
Fixed by commit `2c643dca6a` :heavy_check_mark:

IntegralProgrammer commented:
> 
>     * The indexing servlet returns a 200 status even if there's an error, but with an error payload. The response should be a 500, and the frontend should handle the error payload.
> 
Fixed by commits `95e3fc82` and `b1b927772` :heavy_check_mark:

IntegralProgrammer commented:
> 
>     * Why only `OWL` is supported? The same code can handle all other formats, the only thing that is preventing it now is line 130 where we hadcode the format. Is there a way to detect the format from the input filename?
> 
Fixed by commits `e3577abc`, `4795ec76` where `getSourceFormat()` is moved from `EbiRepositoryHandler` to the utility class `OntologyFormatDetection`. This same function is used here to determine the format of the upload vocabulary file from its file extension. :heavy_check_mark:

sashaandjic commented:
I was able to install HP.obo.
I had to manually enter proper Identifier, Name and Version. Can this be pulled out of the obo file instead?
After uninstall, the screen did not refresh.
![image](https://user-images.githubusercontent.com/23243244/95471292-c8bfa000-094f-11eb-9815-371b11a5e4fc.png)

marta- commented:
> I had to manually enter proper Identifier, Name and Version. Can this be pulled out of the obo file instead?

There should be a `data-version` field in the obo. For HPO it seems to look like this: `data-version: hp/releases/2020-08-11`.

The closest thing I can find to an identifier is `ontology: hp.obo` (for Go the field value is `ontology: go`).

There doesn't seem to be explicit mention of the human readable name of the ontology inside of the obo file.

My suggestion is to try to retrieve the version from the obo, but continue prompt the user for the name. I'm on the fence regarding the id.

sashaandjic commented:
I am able to install the same vocabulary under different ID.
![image](https://user-images.githubusercontent.com/23243244/95476665-76817d80-0955-11eb-9a86-cccb09f5e09b.png)

marta- commented:
> I am able to install the same vocabulary under different ID.

In that case, as an amendment to my previous comment, let's attempt extracting the ID from the obo's `ontology` field and see if that works for most vocabularies. We'll hardcode stripping  down the `.obo` suffix if it appears in the field's value. Without a centralized source like bioportal to keep track of the files it's probably the only easy way to ensure vocabularies are not duplicated.

IntegralProgrammer committed:
Change 'Select OWL / OWL Selected' to 'Select File / File Selected'

IntegralProgrammer committed:
Reset state of vocabulary file upload form upon successful vocabulary upload/installation

sashaandjic commented:
Tested last two commits.
![image](https://user-images.githubusercontent.com/23243244/95595200-e6f1d280-0a19-11eb-8a59-fb084f75be86.png)

IntegralProgrammer committed:
Clean up the layout of the ontology file installer through the use of Material-UI Grid components

IntegralProgrammer committed:
When setting Identifier,Name,Version of a local file ontology, display RED TextField components if the values are empty

IntegralProgrammer committed:
Clean up code styling in owlInstaller.jsx

sashaandjic commented:
Tested. Looking good.
![image](https://user-images.githubusercontent.com/23243244/95628894-efafcc00-0a4c-11eb-8b9b-d8baa4d0c8ea.png)

It may look better like this?
![image](https://user-images.githubusercontent.com/23243244/95628926-00604200-0a4d-11eb-9ab7-cad10eb7476a.png)

marta- commented:
> Would it look better like this?
> _(all in one row; more space between 3 main headers; narrower SELECT FILE button, wider Install button)_

I agree with more space between the headers, but not with having the entire "Install from local file" section on one line. If we were to do any further UI improvements, I suggest:
1. increase the space between sections
2. left-align the local file selection form instead of centering it
3. maybe try `variant="outlined"` for the select button

All of the above are non-blocking and can be done as part of a separate task for polishing the vocabularies UI in general.

sashaandjic commented:
Found one small issue, that will be added to LFS-615. Once vocabulary is installed, the About button is not present on the list of the installed vocabularies as it is for the vocabularies installed via BioPortal. See in the attached screen shot where ORDO was installed via BioPortal unlike HP and HPO.
![image](https://user-images.githubusercontent.com/23243244/95908211-b89e2b00-0d6a-11eb-9a09-c8f4e2cb98e3.png)

IntegralProgrammer committed:
When installing a vocabulary from a local file, disable the Install button until the vocabulary file has been selected and the associated data fields have been filled out

IntegralProgrammer committed:
Remove unnecessary logging

sashaandjic reviewed (approved):
tested - When installing a vocabulary from a local file, disable the Install b…

sashaandjic commented:
One small complaint: After successful install the button needs to go back to disabled state with proper text label "Install"
![image](https://user-images.githubusercontent.com/23243244/96042773-eb642400-0e3b-11eb-84c9-248e38168a35.png)
